micro_sam/evaluation/inference.py

```diff
@@ -392,118 +389,127 @@ def _save_segmentation(masks, prediction_path):
     shape = masks.shape[-2:]
     masks = [{"segmentation": mask, "area": mask.sum()} for mask in masks]
     segmentation = mask_data_to_segmentation(masks, shape, with_background=True)
-    imageio.imwrite(prediction_path, segmentation)
+    imageio.imwrite(prediction_path, segmentation, compression=5)
+
+
+def extract_instances_from_batched_outputs(batched_outputs):
+    masks = [m["segmentation"][None] for m in batched_outputs]
+    masks = torch.stack(masks)
+    masks = (masks > 0.5).to(torch.float32)
+    return masks
+
+
+def get_channelwise_gt_objects(gt, gt_ids, device=None):
+    device = util._get_device(device)
+    sampled_binary_gt = np.stack([(gt == idx)[None] for idx in gt_ids])
+    sampled_binary_gt = torch.from_numpy(sampled_binary_gt).to(torch.float32)
+    sampled_binary_gt = sampled_binary_gt.to(device)
+    return sampled_binary_gt
 
 
+@torch.no_grad()
 def _run_inference_with_iterative_prompting_for_image(
-    model,
+    predictor,
     image,
     gt,
-    n_iterations,
-    device,
+    use_points,
     use_boxes,
-    prediction_paths,
+    dilation,
     batch_size,
-):
-    assert len(prediction_paths) == n_iterations, f"{len(prediction_paths)}, {n_iterations}"
-    to_sam_inputs = ConvertToSamInputs()
+    embedding_path,
+    n_iterations,
+    prediction_paths
+) -> None:
+    prompt_generator = IterativePromptGenerator()
 
-    image = torch.from_numpy(
-        image[None, None] if image.ndim == 2 else image[None]
+    gt_ids = np.unique(gt)[1:]
+    n_positives = 1 if not use_boxes else 0
+    n_negatives = 0
+
+    points, point_labels, boxes = _get_batched_prompts(
+        gt, gt_ids, use_points, use_boxes,
+        n_positives=n_positives,
+        n_negatives=n_negatives,
+        dilation=dilation
     )
-    gt = torch.from_numpy(gt[None].astype("int32"))
 
-    n_pos = 0 if use_boxes else 1
-    batched_inputs, sampled_ids = to_sam_inputs(image, gt, n_pos=n_pos, n_neg=0, get_boxes=use_boxes)
+    sampled_binary_gt = get_channelwise_gt_objects(gt, gt_ids)
 
-    input_images = torch.stack([model.preprocess(x=x["image"].to(device)) for x in batched_inputs], dim=0)
-    image_embeddings = model.image_embeddings_oft(input_images)
-
-    multimasking = n_pos == 1
-    prompt_generator = IterativePromptGenerator()
-
-    n_samples = len(sampled_ids[0])
-    n_batches = int(np.ceil(float(n_samples) / batch_size))
+    # Use multi-masking only if we have a single positive point without box
+    multimasking = False
+    if not use_boxes and (n_positives == 1 and n_negatives == 0):
+        multimasking = True
```

Comment:
You can make the mulitmasking decision above already, it's easier to read this if it's in a single if-else. And you need to always set it to False after the first loop, because we will then have more than one prompt and should not use multimasking anymore.